### PR TITLE
Implement strict directory structure for file provider

### PR DESCRIPTION
**Pull request by snylonue**
probably works now

### CODE REVIEW

File: anni-provider/src/providers/file.rs [hunk 1]
```diff
@@ -17,14 +17,14 @@ pub struct FileBackend {
     repo: Mutex<RepoDatabaseRead>,
     album_path: HashMap<String, PathBuf>,
     album_discs: HashMap<String, Vec<PathBuf>>,
-    strict: bool,
+    strict: (bool, usize), // (enable, layer)
```
Comment (Yesterday17) [inline anni-provider/src/providers/file.rs:20]: `Option<usize>`？

File: anni-provider/src/providers/file.rs [hunk 2]
```diff
@@ -10,21 +10,28 @@ use tokio::fs::{read_dir, File};
...
     album_path: HashMap<String, PathBuf>,
     album_discs: HashMap<String, Vec<PathBuf>>,
+    strict: bool,
 }
 
 impl FileBackend {
-    pub async fn new(root: PathBuf, repo: RepoDatabaseRead) -> Result<Self, ProviderError> {
+    pub async fn new(
+        root: PathBuf,
+        repo: RepoDatabaseRead,
```
[the remark this answers is not in the document]
Comment (snylonue) [reply anni-provider/src/providers/file.rs:26]: > Use `Option<RepoDatabaseRead>` to identify strict/non-strict?

Maybe a enum like `StructreKind`?

**Comment by snylonue** [comment]
改了下 trait `AnniProvider`

File: anni-provider/Cargo.toml [hunk 3]
```diff
@@ -17,6 +17,7 @@ log = "0.4.0"
 
 google-drive3 = { path = "../third_party/google-drive3" }
 
+uuid = { version = "0.8" }
```
Comment (Yesterday17) [inline anni-provider/Cargo.toml:20]: We should use uuid v1 now

File: anni-provider/src/providers/file.rs [hunk 4]
```diff
@@ -128,51 +149,29 @@ impl AnniProvider for FileBackend {
...
+                Some(entry)
+                    if entry
+                        .file_name()
+                        .to_string_lossy()
+                        .starts_with(&format!("{:02}.", track_id)) =>
+                {
+                    break File::open(entry.path()).await?
+                }
+                None => return Err(ProviderError::FileNotFound),
+                _ => {}
```
Comment (Yesterday17) [inline anni-provider/src/providers/file.rs:163]:
```suggestion
```

unreachable?

File: anni-provider/src/providers/file.rs [hunk 5]
```diff
@@ -184,21 +183,147 @@ impl AnniProvider for FileBackend {
...
+    layer: usize,
+}
+
+impl StrictFileBackend {
+    pub fn new(root: PathBuf, layer: usize) -> Self {
+        Self { root, layer }
+    }
+
+    
+    pub fn get_album_path(&self, album_id: &str) -> PathBuf {
```
Comment (Yesterday17) [inline anni-provider/src/providers/file.rs:202]: `get_album_path` -> `album_path`

File: anni-provider/src/providers/file.rs [hunk 6]
```diff
@@ -184,21 +183,147 @@ impl AnniProvider for FileBackend {
...
+            path.join(disc_id.to_string())
+                .join(format!("{track_id}.flac")),
+        ).await?;
+        
+        read_audio(&path, file, range).await
+    }
+
+    async fn get_cover(&self, album_id: &str, disc_id: Option<u8>) -> Result<ResourceReader, ProviderError> {
+        let path = match disc_id {
+            Some(id) => self.get_album_path(album_id).join(id.to_string()),
```
Comment (Yesterday17) [inline anni-provider/src/providers/file.rs:280]:
```suggestion
            Some(disc_id) => self.get_album_path(album_id).join(disc_id.to_string()),
```

**Review by Yesterday17 — CHANGES_REQUESTED** [review]

File: anni-provider/src/providers/file.rs [hunk 7]
```diff
@@ -128,51 +149,29 @@ impl AnniProvider for FileBackend {
     ) -> Result<AudioResourceReader, ProviderError> {
         let path = self.get_disc(album_id, disc_id)?;
         let mut dir = read_dir(path).await?;
-        while let Some(entry) = dir.next_entry().await? {
```
Comment (Yesterday17) [inline anni-provider/src/providers/file.rs:131]: `has_album` should also be implemented for `FileBackend`

File: anni-provider/src/providers/file.rs (hunk 4 above)
Comment (snylonue) [reply anni-provider/src/providers/file.rs:163]: The `Some` branch has a match guard to match `track id`.
The branch is reached when `next_entry()` exists but is not the intended track.

File: anni-provider/src/providers/file.rs (hunk 7 above)
Comment (snylonue) [reply anni-provider/src/providers/file.rs:131]: > `has_album` should also be implemented for `FileBackend`

```
/// Returns whether given album exists
    async fn has_album(&self, album_id: &str) -> bool {
        self.albums().await.unwrap_or(HashSet::new()).contains(album_id)
    }
```
There is a default implementation.

**Review by Yesterday17 — APPROVED** [review]
LGTM

**Pull request merged**